{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "9bcecfd9-0b3b-4875-baca-72a281f5ffea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "multi-matched comp count 3020\n",
      "multi-matched base count 259\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pysam\n",
    "import truvari\n",
    "import json\n",
    "import pandas as pd\n",
    "import joblib\n",
    "import numpy as np\n",
    "\n",
    "SIZEMIN = 50\n",
    "SIZEMAX=10000\n",
    "NEIGHLIMIT = 6\n",
    "program='test'\n",
    "technology='test'\n",
    "# This should have numneigh annotation inside\n",
    "base_vcf = \"dipcall_bcftools_merge_neigh.vcf.gz\"\n",
    "bed_fn = \"dipcall.bed\"\n",
    "comp_vcf = \"sniffles_fixed.vcf.gz\"\n",
    "bench_jl = \"bench_sniffles/data.jl\"\n",
    "bench_summary = \"bench_sniffles/summary.json\"\n",
    "\n",
    "###\n",
    "\n",
    "# Load the data\n",
    "data = joblib.load(bench_jl)\n",
    "\n",
    "# Separate out the variants from the base VCF and add new columns of the base/comp ids\n",
    "base = data[data['state'].isin(['tpbase', 'fn'])].copy()\n",
    "base['base_id'] = base['MatchId'].apply(lambda x: x[0])\n",
    "base['comp_id'] = base['MatchId'].apply(lambda x: x[1])\n",
    "\n",
    "# Separate out the variants from the comparison VCF and add new columns of the base/comp ids\n",
    "comp = data[data['state'].isin(['tp', 'fp'])].copy()\n",
    "comp['base_id'] = comp['MatchId'].apply(lambda x: x[0])\n",
    "comp['comp_id'] = comp['MatchId'].apply(lambda x: x[1])\n",
    "\n",
    "# Merge the base/comparison variants\n",
    "combined = pd.merge(base, comp, left_on='base_id', right_on='comp_id', suffixes=('_base', '_comp'))\n",
    "\n",
    "# How many comp variants matched to multiple base variants?\n",
    "counts1 = combined['base_id_comp'].value_counts()\n",
    "print('multi-matched comp count', (counts1 != 1).sum())\n",
    "\n",
    "# How many base variants matched to multiple comp variants?\n",
    "counts2 = combined['comp_id_base'].value_counts()\n",
    "print('multi-matched base count', (counts2 != 1).sum())\n",
    "\n",
    "# For every tp-base, I can check its GTMatch to figure out concordance/discordance\n",
    "#.. but that doesn't solve the ALstate lookups..\n",
    "# So I need to use a vcf2df \n",
    "# and for every fn, I add that to discorant\n",
    "# and for every fp, I add that to discordant\n",
    "# and then I can add to concordant every reference homozygous base_vcf\n",
    "# And then I can add to missing every Missing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "74b4682b-19e7-42fb-b752-4f544d203882",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_cnt_base = {'Concordant': 0,\n",
    "          'Discordant': 0,\n",
    "          'Missing': 0,\n",
    "          'REFHOM': 0}\n",
    "\n",
    "m_cnt = {'DEL': dict(m_cnt_base),\n",
    "         'INS': dict(m_cnt_base),\n",
    "         'TOT': dict(m_cnt_base)}\n",
    "m_neigh_cnt = {}\n",
    "for i in range(0, NEIGHLIMIT):\n",
    "    m_neigh_cnt[i] = dict(m_cnt_base)\n",
    "\n",
    "# These are relative to the comparison VCF\n",
    "matched = (comp['state'] == 'tp') & (comp['GTMatch'] == 0)\n",
    "m_cnt['TOT']['Concordant'] = matched.sum()\n",
    "m_cnt['TOT']['Discordant'] = (~matched).sum() + (base['state'] == 'fn').sum()\n",
    "\n",
    "sub = comp[comp['svtype'] == 'DEL']\n",
    "matched_del = (sub['state'] == 'tp') & (sub['GTMatch'] == 0)\n",
    "m_cnt['DEL']['Concordant'] = matched_del.sum()\n",
    "m_cnt['DEL']['Discordant'] = (~matched_del).sum() \\\n",
    "                            + ((base['svtype'] == 'DEL') & (base['state'] == 'fp')).sum()\n",
    "\n",
    "sub = comp[comp['svtype'] == 'INS']\n",
    "matched_del = (sub['state'] == 'tp') & (sub['GTMatch'] == 0)\n",
    "m_cnt['INS']['Concordant'] = matched_del.sum()\n",
    "m_cnt['INS']['Discordant'] = (~matched_del).sum() \\\n",
    "                            + ((base['svtype'] == 'INS') & (base['state'] == 'fp')).sum()\n",
    "\n",
    "\n",
    "vcf = pysam.VariantFile(comp_vcf)\n",
    "bed = truvari.build_region_tree(vcfA=vcf, includebed=bed_fn)\n",
    "if 'chrX' in bed:\n",
    "    del(bed['chrX'])\n",
    "if 'chrY' in bed:\n",
    "    del(bed['chrY'])\n",
    "for i in list(bed.keys()):\n",
    "    if '_' in i:\n",
    "        del(bed[i])\n",
    "\n",
    "vcf_i = truvari.region_filter(vcf, bed)\n",
    "\n",
    "for entry in vcf_i:\n",
    "    sz = truvari.entry_size(entry)\n",
    "    if sz < SIZEMIN or sz > SIZEMAX:\n",
    "        continue\n",
    "    m_gt = entry.samples[0]['GT']\n",
    "    svtype = truvari.entry_variant_type(entry).name\n",
    "    # Already accounted for\n",
    "    if 1 in m_gt:\n",
    "        continue\n",
    "    if None in m_gt:\n",
    "        m_cnt['TOT']['Missing'] += 1\n",
    "        m_cnt[svtype]['Missing'] += 1 \n",
    "    else:\n",
    "        m_cnt['TOT']['REFHOM'] += 1\n",
    "        m_cnt[svtype]['REFHOM'] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a57bf618-dc67-4904-824c-a83aff719714",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/fy/lkd5_yz90zq3f67qys3tpztc0000gn/T/ipykernel_73335/205914601.py:16: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  neigh_sum = base.groupby(['NeighBin', 'CorrectGT']).size().unstack()\n"
     ]
    }
   ],
   "source": [
    "parts = []\n",
    "for k,v in m_cnt.items():\n",
    "    d = pd.Series(v)\n",
    "    d['svtype'] = k\n",
    "    parts.append(d.to_frame())\n",
    "summary = pd.concat(parts, axis=1)\n",
    "summary = summary.T.reset_index(drop=True)\n",
    "summary['GT Concordance'] = summary['Concordant'] / (summary['Concordant'] + summary['Discordant'])\n",
    "\n",
    "\n",
    "base['NeighBin'] = pd.cut(base['NumNeighbors'], bins=[0, 1, 2, 3, 4, 5, sys.maxsize],\n",
    "                            labels=['0', '1', '2', '3', '4', '>=5'])\n",
    "       \n",
    "base['CorrectGT'] = (base['state'] == 'tpbase') & (base['GTMatch'] == 0)\n",
    "\n",
    "neigh_sum = base.groupby(['NeighBin', 'CorrectGT']).size().unstack()\n",
    "neigh_sum['pct'] = neigh_sum[True] / neigh_sum.sum(axis=1)\n",
    "neigh_sum['total'] = neigh_sum[True] + neigh_sum[False]\n",
    "neigh_sum['pct of baseline']  = neigh_sum['total'] / neigh_sum['total'].sum()\n",
    "\n",
    "perf_sum = json.load(open(bench_summary))\n",
    "del(perf_sum['gt_matrix'])\n",
    "perf_sum['Baseline GT Concordance'] = perf_sum['gt_concordance']\n",
    "del(perf_sum['gt_concordance'])\n",
    "perf_sum['Concordant'] = perf_sum['TP-base_TP-gt']\n",
    "perf_sum['GT Accuracy'] = perf_sum['TP-base_TP-gt'] / perf_sum['base cnt']\n",
    "del(perf_sum['TP-comp_TP-gt'])\n",
    "del(perf_sum['TP-comp_FP-gt'])\n",
    "del(perf_sum['TP-base_TP-gt'])\n",
    "del(perf_sum['TP-base_FP-gt'])\n",
    "\n",
    "perf_sum = pd.DataFrame([perf_sum])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a49a4517-dd74-4353-8b8d-ca0ee3394b74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([168049.,  97753., 103178.,   2054.])"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtc = comp['GT'].apply(truvari.get_gt).value_counts()\n",
    "\n",
    "m_gtarray = np.zeros(4)\n",
    "m_gtarray[0] = m_cnt['TOT'][\"REFHOM\"]\n",
    "m_gtarray[1] = gtc[truvari.GT.HET]\n",
    "m_gtarray[2] = gtc[truvari.GT.HOM]\n",
    "m_gtarray[3] = m_cnt['TOT']['Missing']\n",
    "m_gtarray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1f5cb433-4f79-45e9-9f83-89f2bf5baaaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TP-base</th>\n",
       "      <th>TP-comp</th>\n",
       "      <th>FP</th>\n",
       "      <th>FN</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1</th>\n",
       "      <th>base cnt</th>\n",
       "      <th>comp cnt</th>\n",
       "      <th>baseline gt_concordance</th>\n",
       "      <th>Concordant</th>\n",
       "      <th>GT Concordant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23502</td>\n",
       "      <td>23510</td>\n",
       "      <td>177421</td>\n",
       "      <td>1987</td>\n",
       "      <td>0.117005</td>\n",
       "      <td>0.922045</td>\n",
       "      <td>0.207659</td>\n",
       "      <td>25489</td>\n",
       "      <td>200931</td>\n",
       "      <td>0.71829</td>\n",
       "      <td>16886</td>\n",
       "      <td>0.662482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   TP-base  TP-comp      FP    FN  precision    recall        f1  base cnt  \\\n",
       "0    23502    23510  177421  1987   0.117005  0.922045  0.207659     25489   \n",
       "\n",
       "   comp cnt  baseline gt_concordance  Concordant  GT Concordant  \n",
       "0    200931                  0.71829       16886       0.662482  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Matching the non-reference-homozygous variants between the baseline multi-sample VCF\n",
    "# and the genotyper results\n",
    "# Recall is how many baseline variants are still present in the genotyper results\n",
    "# Precision is how many comparison variants are present and have â‰¥90% similarity to a present baseline variant\n",
    "# baseline gt_concordance is the percent of baseline variants with a correct genotype\n",
    "perf_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2a1d95bf-70be-4acd-a5f7-1c2330c52e86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Concordant</th>\n",
       "      <th>Discordant</th>\n",
       "      <th>Missing</th>\n",
       "      <th>REFHOM</th>\n",
       "      <th>svtype</th>\n",
       "      <th>GT Concordance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7529</td>\n",
       "      <td>31879</td>\n",
       "      <td>460</td>\n",
       "      <td>50298</td>\n",
       "      <td>DEL</td>\n",
       "      <td>0.191053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9358</td>\n",
       "      <td>152165</td>\n",
       "      <td>1594</td>\n",
       "      <td>117751</td>\n",
       "      <td>INS</td>\n",
       "      <td>0.057936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16887</td>\n",
       "      <td>186031</td>\n",
       "      <td>2054</td>\n",
       "      <td>168049</td>\n",
       "      <td>TOT</td>\n",
       "      <td>0.083221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Concordant Discordant Missing  REFHOM svtype GT Concordance\n",
       "0       7529      31879     460   50298    DEL       0.191053\n",
       "1       9358     152165    1594  117751    INS       0.057936\n",
       "2      16887     186031    2054  168049    TOT       0.083221"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For the comparison variants, we can count how many genotypes are concordant as those are present\n",
    "# And match to a present baseline variant. We can also count the number of reference homozygous variants\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ce4ff336-faae-4a04-a33e-0aff03e79b9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>CorrectGT</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "      <th>pct</th>\n",
       "      <th>total</th>\n",
       "      <th>pct of baseline</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NeighBin</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>171</td>\n",
       "      <td>1175</td>\n",
       "      <td>0.8730</td>\n",
       "      <td>1346</td>\n",
       "      <td>0.0628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>210</td>\n",
       "      <td>1097</td>\n",
       "      <td>0.8393</td>\n",
       "      <td>1307</td>\n",
       "      <td>0.0610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>223</td>\n",
       "      <td>831</td>\n",
       "      <td>0.7884</td>\n",
       "      <td>1054</td>\n",
       "      <td>0.0492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>253</td>\n",
       "      <td>669</td>\n",
       "      <td>0.7256</td>\n",
       "      <td>922</td>\n",
       "      <td>0.0430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>210</td>\n",
       "      <td>575</td>\n",
       "      <td>0.7325</td>\n",
       "      <td>785</td>\n",
       "      <td>0.0366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;=5</th>\n",
       "      <td>7355</td>\n",
       "      <td>8660</td>\n",
       "      <td>0.5407</td>\n",
       "      <td>16015</td>\n",
       "      <td>0.7474</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CorrectGT  False  True     pct  total  pct of baseline\n",
       "NeighBin                                              \n",
       "0            171  1175  0.8730   1346           0.0628\n",
       "1            210  1097  0.8393   1307           0.0610\n",
       "2            223   831  0.7884   1054           0.0492\n",
       "3            253   669  0.7256    922           0.0430\n",
       "4            210   575  0.7325    785           0.0366\n",
       ">=5         7355  8660  0.5407  16015           0.7474"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For the baseline variants, we count what percent of the present variants are by number of neighbors\n",
    "neigh_sum.round(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aafdb88a-cc01-4497-af38-763c410305d1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
